File: leetcode/algorithms/637_average_of_levels_in_binary_tree/main.cpp
#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class AverageOfLevelsInBinaryTree {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> result;
        vector<TreeNode*> queue(10000);
        int left = 0, right = 0;

        queue[right++] = root;
        while (left < right) {
            int nodeCount = right - left;
            double sum = 0;
            for (int i = 0; i < nodeCount; i++) {
                TreeNode* node = queue[left++];
                sum += node->val;
                if (node->left) {
                    queue[right++] = node->left;
                }
                if (node->right) {
                    queue[right++] = node->right;
                }
            }
            result.push_back(sum / nodeCount);
        }

        return result;
    }

    // Solution
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            long temp = 0;
            int s = q.size();
            for (int i = 0; i < s; i++) {
                TreeNode* t = q.front();
                q.pop();
                if (t->left) {
                    q.push(t->left);
                }
                if (t->right) {
                    q.push(t->right);
                }
                temp += t->val;
            }
            res.push_back((double)temp / s);
        }
        return res;
    }
};

File: leetcode/algorithms/637_average_of_levels_in_binary_tree/main.dart
import 'dart:collection';

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;
  TreeNode([this.val = 0, this.left, this.right]);
}

class AverageOfLevelsInBinaryTree {
  List<double> averageOfLevels(TreeNode? root) {
    final List<double> result = List<double>.empty(growable: true);
    final Queue<TreeNode> nodeQueue = Queue<TreeNode>();

    if (root != null) {
      nodeQueue.add(root);
      while (nodeQueue.isNotEmpty) {
        int nodeCount = nodeQueue.length;
        double sum = 0;
        for (int i = 0; i < nodeCount; i++) {
          TreeNode node = nodeQueue.removeFirst();
          sum += node.val;
          if (node.left != null) {
            nodeQueue.add(node.left!);
          }
          if (node.right != null) {
            nodeQueue.add(node.right!);
          }
        }
        result.add(sum / nodeCount);
      }
    }

    return result;
  }

  // Solution
  List<double> solution(TreeNode? root) {
    if (root == null) {
      return [];
    }
    final queue = Queue<TreeNode>();
    queue.add(root);
    List<double> sol = [];
    while (queue.isNotEmpty) {
      int n = queue.length;
      int sum = 0;
      for (int i = 0; i < n; i++) {
        TreeNode node = queue.removeFirst();
        sum += node.val;
        if (node.left != null) {
          queue.add(node.left!);
        }
        if (node.right != null) {
          queue.add(node.right!);
        }
      }
      sol.add(sum / n);
    }
    return sol;
  }
}
